File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.BLL/Action/Action1442.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;
using System.Data;
using ZyGames.Tianjiexing.BLL.Base;
using ZyGames.Tianjiexing.Component;
using ZyGames.Tianjiexing.Lang;
using ZyGames.Tianjiexing.Model;
using ZyGames.Framework.Game.Service;
using ZyGames.Framework.Game.Cache;
using ZyGames.Framework.Common;

namespace ZyGames.Tianjiexing.BLL.Action
{

    /// <summary>
    /// 1442_佣兵升级接口
    /// </summary>
    public class Action1442 : BaseAction
    {
        private int generalID;

        public Action1442(ZyGames.Framework.Game.Contract.HttpGet httpGet)
            : base(ActionIDDefine.Cst_Action1442, httpGet)
        {

        }

        public override void BuildPacket()
        {

        }

        public override bool GetUrlElement()
        {
            if (httpGet.GetInt("GeneralID", ref generalID))
            {
                return true;
            }
            return false;
        }

        public override bool TakeAction()
        {
            var package = UserItemPackage.Get(ContextUser.UserID);
            UserGeneral userGeneral = new GameDataCacheSet<UserGeneral>().FindKey(ContextUser.UserID, generalID);
            GeneralInfo general = new ConfigCacheSet<GeneralInfo>().FindKey(generalID);
            if (package == null || general == null)
            {
                ErrorCode = LanguageManager.GetLang().ErrorCode;
                return false;
            }
            if (userGeneral == null)
            {
                ErrorCode = LanguageManager.GetLang().ErrorCode;
                ErrorInfo = LanguageManager.GetLang().St1442_SelectMercenaryUpgrade;
                return false;
            }
            if (string.IsNullOrEmpty(userGeneral.GeneralCard))
            {
                ErrorCode = LanguageManager.GetLang().ErrorCode;
                ErrorInfo = LanguageManager.GetLang().St1442_SelectTheExperienceCard;
                return false;
            }
            string[] itemStr = userGeneral.GeneralCard.Split(',');
            bool isUse = false;
            foreach (var str in itemStr)
            {
                var useritem = package.ItemPackage.Find(s => !s.IsRemove && s.UserItemID == str);
                if (useritem == null)
                {
                    continue;
                }
                int exprience = 0;
                ItemBaseInfo itemInfo = new ConfigCacheSet<ItemBaseInfo>().FindKey(useritem.ItemID);
                if (itemInfo != null)
                {
                    exprience = itemInfo.EffectNum;
                    if (general.ItemID == itemInfo.ItemID)
                    {
                        exprience = (exprience * GameConfigSet.ExpMultiple).ToInt();
                    }
                    //GeneralHelper.UserGeneralExp(ContextUser.UserID, generalID, exprience);
                    UserItemHelper.UseUserItem(ContextUser.UserID, useritem.ItemID, 1);
                    isUse = true;
                }
            }
            if (isUse)
            {
                userGeneral.GeneralCard = string.Empty;
            }
            return true;
        }
    }
}

File: Sample/Koudai/Server/release/Script/PyScript/Action/action1442.py
import clr, sys
clr.AddReference('ZyGames.Framework.Common');
clr.AddReference('ZyGames.Framework');
clr.AddReference('ZyGames.Framework.Game');
clr.AddReference('ZyGames.Tianjiexing.Model');
clr.AddReference('ZyGames.Tianjiexing.BLL');
clr.AddReference('ZyGames.Tianjiexing.Lang');
clr.AddReference('ZyGames.Tianjiexing.Component');
from action import *
from System import *
from lang import Lang
from System.Collections.Generic import *
from ZyGames.Framework.Common.Log import *
from ZyGames.Tianjiexing.Model.ConfigModel import *
from ZyGames.Framework.Common import *
from ZyGames.Tianjiexing.Model import *
from ZyGames.Tianjiexing.BLL import *
from ZyGames.Tianjiexing.Lang import *
from ZyGames.Framework.Game.Cache import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Tianjiexing.Model.Config import *
from ZyGames.Tianjiexing.Model.Enum import *
from ZyGames.Tianjiexing.BLL.Base import *
from ZyGames.Tianjiexing.Component import *
from ZyGames.Framework.Common.Serialization import *

# 1442_佣兵升级接口
class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)
        self.generalID = 0

class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)

def getUrlElement(httpGet, parent):
    urlParam = UrlParam();
    if httpGet.Contains("GeneralID"):
        urlParam.generalID = httpGet.GetIntValue("GeneralID");
    else:
        urlParam.Result = False;
    return urlParam;

def takeAction(urlParam, parent):
    actionResult = ActionResult();
    userId = parent.Current.User.PersonalId;
    contextUser = parent.Current.User;
   
    package = UserItemPackage.Get(userId);
    userGeneral = GameDataCacheSet[UserGeneral]().FindKey(userId, urlParam.generalID);
    general = ConfigCacheSet[GeneralInfo]().FindKey(urlParam.generalID);
    if package == None or general == None:
        parent.ErrorCode = Lang.getLang("ErrorCode");
        parent.ErrorInfo = Lang.getLang("LoadError");
        actionResult.Result = False;
        return actionResult;
    if userGeneral == None:
        parent.ErrorCode = Lang.getLang("ErrorCode");
        parent.ErrorInfo = Lang.getLang("LoadError");
        actionResult.Result = False;
        return actionResult;
    if not userGeneral.GeneralCard:
        parent.ErrorCode = LanguageManager.GetLang().ErrorCode;
        parent.ErrorInfo = LanguageManager.GetLang().St1442_SelectTheExperienceCard;
        actionResult.Result = False;
        return actionResult;
    if userGeneral.GeneralLv >= (contextUser.UserLv*3):
       parent.ErrorCode = Lang.getLang("ErrorCode");
       parent.ErrorInfo = Lang.getLang("St1442_GeneralLvNotUserLv");
       actionResult.Result = False;
       return actionResult;
    itemStr = userGeneral.GeneralCard.Split(',');
    itemNum=userGeneral.GeneralCardNum.Split(',');   #获取佣兵经验卡的数量
    isUse = False;
    #总的经验
    i=0               #用于循环到第几个佣兵
    for str in itemStr:
        exprience = 0; 
        useritem = package.ItemPackage.Find(lambda s:not s.IsRemove and s.UserItemID == str);
        if useritem :
            itemInfo = ConfigCacheSet[ItemBaseInfo]().FindKey(useritem.ItemID);  
            if itemInfo != None:
                conunt = itemNum[i]
                conunt =MathUtils.ToInt(conunt)
                for expri in range(conunt):
                    exprience = MathUtils.ToInt(itemInfo.EffectNum);    #经验多张佣兵卡
                    if general.ItemID == itemInfo.ItemID:
                        exprience = MathUtils.Addition(exprience,exprience*GameConfigSet.ExpMultiple);   #经验多张佣兵卡
                    # 判断佣兵等级是否达到上限
                    status = UserHelper.GeneralLvIsUserLv(userId,urlParam.generalID,exprience,contextUser.UserLv);
                    if status == 0:
                        GeneralHelper.UserGeneralExp(userId, urlParam.generalID, exprience);  # 佣兵加经验
                        UserItemHelper.UseUserItem(userId, useritem.ItemID, 1);  # 使用物品
                    isUse = True;
                    #if status == 2:
                    #    parent.ErrorCode = Lang.getLang("ErrorCode");
                    #    parent.ErrorInfo = Lang.getLang("St1442_GeneralLvNotUserLv");
                    #    actionResult.Result = False;
                    #    return actionResult;
                    if status == 3:
                        parent.ErrorCode = Lang.getLang("ErrorCode");
                        parent.ErrorInfo = Lang.getLang("LoadError");
                        actionResult.Result = False;
                        return actionResult;
                    if status == 1:
                        parent.ErrorCode = Lang.getLang("ErrorCode");
                        parent.ErrorInfo = Lang.getLang("St1442_GeneralLvIsMax");
                        actionResult.Result = False;
                        return actionResult;
                    if isUse:
                        userGeneral.GeneralCard = '';
        i+=1  
    userGeneral.RefreshMaxLife();
    return actionResult;

def buildPacket(writer, urlParam, actionResult):
    
    return True;